tests/acceptance/features/bootstrap/TwoFactorTOTPContext.php

```diff
@@ -49,10 +73,71 @@ public function __construct(
 
 	/**
 	 * @Given /^the user has activated TOTP Second\-factor auth but not verified$/
+	 * @When /^the user activates TOTP Second\-factor auth but does not verify$/
 	 *
 	 * @return void
 	 */
 	public function theUserHasActivatedTOTPSecondFactorAuthButNotVerified() {
 		$this->personalSecuritySettingsPage->activateTOTP();
+		$this->totpSecret = $this->personalSecuritySettingsPage->getSecretCode();
+	}
+
+	/**
+	 * Returns secret code extracted from QRCode
+	 *
+	 * @return string
+	 */
+	public function getSecretCodeFromQRCode() {
+		$path = \tempnam(\sys_get_temp_dir(), 'totp_qrcode');
+		$data = \explode(',', $this->personalSecuritySettingsPage->getQRCode());
+		$file = \fopen($path, 'wb');
+		\fwrite($file, \base64_decode($data[1]));
+		$qrCode = new QrReader($path);
+		\preg_match('/secret=(?P<secret>[A-Z0-9]{16})/', $qrCode->text(), $matches);
+		return $matches['secret'];
+	}
+
+	/**
+	 * @When the user adds one-time key generated from the secret key using the webUI
+	 *
+	 * @return void
+	 */
+	public function theUserAddsVerficationKeyFromSecretKeyUsingWebUI() {
+		$remainder = \time() % 30;
+		if ($remainder % 30 < 10) {
+			// When we reach the step to add $key, the key might have already changed.
+			// Let's sleep for a while and then get the key and fill.
+			// Hopefully, it will be completed within 10 seconds.
+			$message = "INFORMATION: Hibernating for $remainder seconds.";
+			echo $message;
+			\error_log($message);
+			\sleep($remainder);
```

Comment:
Why? For example, if ``remainder`` is ``7`` then we are 7 seconds into a 30-second period. Then we wait another 7 seconds, so we are now 14 seconds into the 30 second period.

Maybe you intend to wait ``30 - $remainder + 1`` seconds, to get to the start of the next 30-second period, or?

What is the objective here?